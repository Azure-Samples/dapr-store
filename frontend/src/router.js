import VueRouter from 'vue-router'

import ProductCatalog from './components/ProductCatalog'
import ProductOffers from './components/ProductOffers'
import Home from './components/Home'

const router = new VueRouter({
  routes: [
    {
      path: '/catalog',
      component: ProductCatalog
    },
    {
      path: '/offers',
      component: ProductOffers
    },
    {
      path: '/',
      component: Home
    }
  ]
})

export default router