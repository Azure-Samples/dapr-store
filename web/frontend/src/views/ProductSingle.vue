<!--
// ----------------------------------------------------------------------------
// Copyright (c) Ben Coleman, 2020
// Licensed under the MIT License.
//
// Dapr Store frontend - Show a single product
// ----------------------------------------------------------------------------
-->

<template>
  <div>
    <h1>Product Details</h1>
    <error-box :error="error" />

    <div v-if="!product && !error" class="text-center">
      <b-spinner variant="success" style="width: 5rem; height: 5rem;" />
    </div>

    <b-card v-if="product">
      <b-row class="d-flex">
        <b-col class="mb-3">
          <b-card-title>
            {{ product.name }}
          </b-card-title>
          <div v-if="product.onOffer" class="onsale">
            On Sale
          </div>
          <br><br><br>
          {{ product.description }}
          <br><br>
          Â£{{ product.cost }}
          <br><br>
          <b-button id="addBut" :disabled="!isLoggedIn()" variant="primary" @click="addToCart">
            <fa icon="shopping-cart" />
            &nbsp; Add to Cart
          </b-button>
        </b-col>

        <b-col>
          <div class="product-img">
            <img :src="product.image">
          </div>
        </b-col>
      </b-row>
    </b-card>
  </div>
</template>

<script>
import ErrorBox from '../components/ErrorBox'
import api from '../services/api'
import auth from '../services/auth'

export default {
  name: 'ProductSingle',

  components: {
    'error-box': ErrorBox
  },

  data() {
    return {
      product: null,
      error: null
    }
  },

  async mounted() {
    try {
      this.product = await api.productGet(this.$route.params.id)
    } catch (err) {
      this.error = err
    }
  },

  methods: {
    async addToCart() {
      try {
        const user = auth.user()
        if (!user) { return }

        await api.cartAddAmount(user.userName, this.product.id, +1)
        this.showToast('Added to your cart!', 'success')
      } catch (err) {
        this.showToast('Error adding to cart ðŸ˜« '+err.toString(), 'danger')
      }
    },

    isLoggedIn() {
      if (auth.user()) { return true }
      return false
    },

    showToast(msg, variant) {
      this.$bvToast.toast(`${this.product.name}`, {
        title: msg,
        variant: variant,
        autoHideDelay: 3000,
        appendToast: true,
        toaster: 'b-toaster-top-center',
        solid: true
      })
    }
  }
}
</script>

<style scoped>
.card-title {
  font-size: 2.5rem;
}
.card-body {
  font-size: 1.1rem;
}
.product-img img {
  width: 100%;
  max-width: 600px;
  border-radius: 1vw;
}
@media  (max-width: 768px) {
  .row {
    flex-direction: column;
  }
}
.onsale {
  display: inline-block;
  width: 100%;
  height: 2rem;
  font-size: 1.2rem;
  line-height: 2rem;
  color: rgb(97, 9, 9);
  background-color: rgb(240, 216, 216);
  text-align: center;
  border-radius: 0.3rem;
}
</style>