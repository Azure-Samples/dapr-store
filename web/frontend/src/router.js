// ----------------------------------------------------------------------------
// Copyright (c) Ben Coleman, 2020
// Licensed under the MIT License.
//
// Dapr Store frontend - client side routes
// ----------------------------------------------------------------------------

import { createRouter, createWebHistory } from 'vue-router'

import auth from './services/auth'

import ProductCatalog from './views/ProductCatalog.vue'
import ProductOffers from './views/ProductOffers.vue'
import ProductSearch from './views/ProductSearch.vue'
import ProductSingle from './views/ProductSingle.vue'
import ViewOrder from './views/ViewOrder.vue'
import Login from './views/Login.vue'
import Account from './views/Account.vue'
import Home from './views/Home.vue'
import About from './views/About.vue'
import Cart from './views/Cart.vue'

const router = createRouter({
  history: createWebHistory(),
  routes: [
    {
      name: 'home',
      path: '/',
      component: Home
    },
    {
      name: 'login',
      path: '/login',
      component: Login
    },
    {
      name: 'account',
      path: '/account',
      component: Account,
      beforeEnter: signedInCheck
    },
    {
      name: 'cart',
      path: '/cart',
      component: Cart,
      beforeEnter: signedInCheck
    },
    {
      name: 'about',
      path: '/about',
      component: About
    },
    {
      name: 'search',
      path: '/search/:query',
      component: ProductSearch
    },
    {
      name: 'catalog',
      path: '/catalog',
      component: ProductCatalog
    },
    {
      name: 'offers',
      path: '/offers',
      component: ProductOffers
    },
    {
      name: 'single-product',
      path: '/product/:id',
      component: ProductSingle
    },
    {
      name: 'view-order',
      path: '/order/:id',
      component: ViewOrder
      //beforeEnter: signedInCheck
    }
  ]
})

function signedInCheck(to, from, next) {
  const user = auth.user()
  if (!user || !user.username) {
    next('/login')
  } else {
    next()
  }
}

export default router
