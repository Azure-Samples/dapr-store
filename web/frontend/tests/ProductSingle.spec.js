import { describe, it, vi, expect } from 'vitest'
import { mount, flushPromises } from '@vue/test-utils'
import router from '@/router'

import ProductSingle from '@/views/ProductSingle.vue'

const productId = 'prd1'
vi.mock('@/services/api')

describe('ProductSingle.vue', () => {
  it('renders product details', async () => {
    router.push('/product/' + productId)
    await router.isReady()

    const wrapper = mount(ProductSingle, {
      global: {
        plugins: [router]
      }
    })

    await flushPromises()
    expect(wrapper.html()).toMatchSnapshot()
  })
})
