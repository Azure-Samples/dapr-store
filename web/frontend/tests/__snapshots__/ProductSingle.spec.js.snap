// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductSingle.vue > renders product details 1`] = `
"<div data-v-fdd171a3="">
  <h1 data-v-fdd171a3="">Product Details</h1>
  <!--v-if-->
  <!--v-if-->
  <div data-v-fdd171a3="" class="card">
    <div data-v-fdd171a3="" class="card-body">
      <div data-v-fdd171a3="" class="row d-flex">
        <div data-v-fdd171a3="" class="col mb-3">
          <div data-v-fdd171a3="" class="card-title">Top Hat (6″)</div>
          <!--v-if-->
          <div data-v-fdd171a3="" class="m-3">Made from 100% wool and quite nice</div>
          <h3 data-v-fdd171a3="">£39.95</h3><button data-v-fdd171a3="" id="addBut" class="btn btn-primary" disabled=""><i data-v-fdd171a3="" class="fa-solid fa-basket-shopping"></i> &nbsp; Add to Cart </button>
        </div>
        <div data-v-fdd171a3="" class="col">
          <div data-v-fdd171a3="" class="product-img"><img data-v-fdd171a3="" src="/img/catalog/1.jpg"></div>
        </div>
      </div>
    </div>
  </div>
  <div data-v-fdd171a3="" class="position-fixed top-0 start-50 translate-middle-x p-3" style="z-index: 11;">
    <div data-v-fdd171a3="" class="toast hide" role="alert">
      <div data-v-fdd171a3="" class="d-flex">
        <div data-v-fdd171a3="" class="toast-body fs-4">Top Hat (6″) was added to your cart!</div><button data-v-fdd171a3="" type="button" class="btn-close me-2 m-auto" data-bs-dismiss="toast" aria-label="Close"></button>
      </div>
    </div>
  </div>
</div>"
`;
